### PR TITLE
Decompile alpha_puzzle overlay

### PR DESCRIPTION
**Pull request by adrienntindall**

### CODE REVIEW

File: include/alph_puzzle.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    u8 unk94[0x40];
+    NNSG2dScreenData *screenData;
+    void *unkD8;
+    AlphPuzzleTile tileGrid[16];
+    AlphPuzzleTile *selectedTile;
+} AlphPuzzleData;
+
+BOOL ov110_AlphPuzzle_OvyInit(OVY_MANAGER *man, int *state);
+BOOL ov110_AlphPuzzle_OvyExec(OVY_MANAGER *man, int *state);
+BOOL ov110_AlphPuzzle_OvyExit(OVY_MANAGER *man);
```
Comment (tgsm) [inline include/alph_puzzle.h:73]: You still need the state param, even if it's not used

File: src/alph_puzzle.c [hunk 2]
```diff
@@ -0,0 +1,971 @@
...
+typedef struct UnkStruct_021E6D20 {
+    AlphPuzzleData *data;
+    u32 unk4;
+} UnkStruct_021E6D20;
+
+static void ov110_021E6D20(AlphPuzzleData *data) {
+    UnkStruct_021E6D20 *unkStruct = AllocFromHeapAtEnd(data->heapId, sizeof(UnkStruct_021E6D20));
+    MI_CpuFill8(unkStruct, 0, sizeof(AlphPuzzleTile));
+    unkStruct->data = data;
+    CreateSysTask(ov110_021E6D54, unkStruct, 0);
```
Comment (tgsm) [inline src/alph_puzzle.c:957]: You can use `UnkStruct_021E6D20` for the task function, you just have to cast the function to a `SysTaskFunc` when calling `CreateSysTask`:
```suggestion
    CreateSysTask((SysTaskFunc)ov110_021E6D54, unkStruct, 0);
```

File: src/alph_puzzle.c [hunk 3]
```diff
@@ -0,0 +1,971 @@
...
+    ov110_021E6904(data, 1);
+    data->unk1A = 1;
+}
+
+static void ov110_021E6D54(SysTask *task, void *_data) {
```
Comment (tgsm) [inline src/alph_puzzle.c:962]:
```suggestion
static void ov110_021E6D54(SysTask *task, UnkStruct_021E6D20 *data) {
```

File: src/alph_puzzle.c [hunk 4]
```diff
@@ -0,0 +1,971 @@
...
+}
+
+extern UnkStruct_0200D2B4 ov110_021E6F7C[2];
+extern UnkStruct_0200D2B4 ov110_021E6FCC;
+
+static void ov110_021E6764(AlphPuzzleData *data) {
+    int i;
+    for (i = 0; i <= 1; i++) {
+        data->unk8C[i] = SpriteRenderer_CreateSprite(data->unk84, data->unk88, &ov110_021E6F7C[i]);
+        sub_02024868(data->unk8C[i], 1 << 12);
```
Comment (tgsm) [inline src/alph_puzzle.c:776]:
```suggestion
        sub_02024868(data->unk8C[i], FX32_ONE);
```

File: src/alph_puzzle.c [hunk 5]
```diff
@@ -0,0 +1,971 @@
...
+    PaletteData_FreeBuffers(data->palette, 2);
+    PaletteData_FreeBuffers(data->palette, 1);
+    PaletteData_FreeBuffers(data->palette, 0);
+    PaletteData_Free(data->palette);
+}
+
+static void ov110_021E6580(AlphPuzzleData *data) {
+    FontID_Alloc(4, data->heapId);
+    
+    data->msgData = NewMsgDataFromNarc(MSGDATA_LOAD_DIRECT, NARC_msgdata_msg, 2, data->heapId);
```
Comment (tgsm) [inline src/alph_puzzle.c:691]:
```suggestion
    data->msgData = NewMsgDataFromNarc(MSGDATA_LOAD_DIRECT, NARC_msgdata_msg, NARC_msg_msg_0002_bin, data->heapId);
```

File: src/alph_puzzle.c [hunk 6]
```diff
@@ -0,0 +1,971 @@
...
+    data->messageFormat = MessageFormat_New_Custom(6, 16, data->heapId);
+    data->unk30 = String_New(0x80, data->heapId);
+    
+    data->unk34 = NewString_ReadMsgData(data->msgData, 0);
```
Comment (tgsm) [inline src/alph_puzzle.c:695]: (include msgdata/msg/msg_0002.h)
```suggestion
    data->unk34 = NewString_ReadMsgData(data->msgData, msg_0002_00000);
```

File: src/alph_puzzle.c [hunk 7]
```diff
@@ -0,0 +1,971 @@
...
+    
+    for (int i = 0; i < 4; i++) {
+        data->unk38[i] = NewString_ReadMsgData(data->msgData, i + 1);
```
Comment (tgsm) [inline src/alph_puzzle.c:698]:
```suggestion
        data->unk38[i] = NewString_ReadMsgData(data->msgData, i + msg_0002_00001);
```

File: src/alph_puzzle.c [hunk 8]
```diff
@@ -0,0 +1,971 @@
...
+    
+    u16 *unkPtr = (u16 *)(0x04000304);
+    *unkPtr |= (1 << 15);
+}
+
+static void ov110_021E6394(AlphPuzzleData *data) {
+    NARC *narc = NARC_New(NARC_a_1_7_2, data->heapId);
+    data->palette = PaletteData_Init(data->heapId);
+    
+    PaletteData_AllocBuffers(data->palette, 0, (1 << 8), data->heapId);
```
Comment (tgsm) [inline src/alph_puzzle.c:654]:
```suggestion
    PaletteData_AllocBuffers(data->palette, 0, 256, data->heapId);
```

File: src/alph_puzzle.c [hunk 9]
```diff
@@ -0,0 +1,971 @@
...
+    PaletteData_AllocBuffers(data->palette, 1, 256, data->heapId);
+    PaletteData_AllocBuffers(data->palette, 2, 256, data->heapId);
+    
+    sub_02003220(data->palette, narc, 10, data->heapId, 0, (1 << 8), 0, 0);
```
Comment (tgsm) [inline src/alph_puzzle.c:658]:
```suggestion
    sub_02003220(data->palette, narc, 10, data->heapId, 0, 256, 0, 0);
```

File: src/alph_puzzle.c [hunk 10]
```diff
@@ -0,0 +1,971 @@
...
+    FreeBgTilemapBuffer(data->bgConfig, 3);
+    FreeBgTilemapBuffer(data->bgConfig, 2);
+    FreeBgTilemapBuffer(data->bgConfig, 0);
+    FreeBgTilemapBuffer(data->bgConfig, 7);
+    FreeBgTilemapBuffer(data->bgConfig, 6);
+    FreeBgTilemapBuffer(data->bgConfig, 4);
+    FreeToHeap(data->bgConfig);
+    
+    u16 *unkPtr = (u16 *)(0x04000304);
+    *unkPtr |= (1 << 15);
```
Comment (tgsm) [inline src/alph_puzzle.c:647]:
```suggestion
    GX_SetDispSelect(GX_DISP_SELECT_MAIN_SUB);
```

File: src/alph_puzzle.c [hunk 11]
```diff
@@ -0,0 +1,971 @@
...
+    u16 *unkPtr;
+    ov110_021E61B0();
+    data->bgConfig = BgConfig_Alloc(data->heapId);
+
+    GraphicsModes mode = ov110_021E6DC0;
+    
+    SetBothScreensModesAndDisable(&mode);
+
+    unkPtr = (u16 *)(0x04000304);
+    *unkPtr = *unkPtr & ~(1 << 15);
```
Comment (tgsm) [inline src/alph_puzzle.c:605]: GX_SetDispSelect(GX_DISP_SELECT_SUB_MAIN);

File: src/alph_puzzle.c [hunk 12]
```diff
@@ -0,0 +1,971 @@
...
+    if (data->unk84) {
+        thunk_OamManager_ApplyAndResetBuffers();
+    }
+
+    NNS_GfdDoVramTransfer();
+    DoScheduledBgGpuUpdates(data->bgConfig);
+
+    //This is probably an inline somewhere but I couldn't find it
+    u32 *ptr = (u32 *) 0x27E0000;
+    ptr[0xFFE] |= 1;
```
Comment (tgsm) [inline src/alph_puzzle.c:545]:
```suggestion
    OS_SetIrqCheckFlag(OS_IE_VBLANK);
```

File: src/alph_puzzle.c [hunk 13]
```diff
@@ -0,0 +1,971 @@
...
+
+static int ov110_021E5CCC(AlphPuzzleData *data) {
+    if (!System_GetTouchHeld()) {
+        data->unkE = 0;
+        return ALPH_PUZZLE_STATE_ROTATE_TILE;
+    }
+    if (data->unkE++ >= 2) {
+        data->unkE = 0;
+        data->unk1C = (data->selectedTile->x << 5) + 0x40;
+        data->unk1E = (data->selectedTile->y << 5) + 0x20;
```
Comment (tgsm) [inline src/alph_puzzle.c:323]:
```suggestion
        data->unk1C = (data->selectedTile->x * 32) + 0x64;
        data->unk1E = (data->selectedTile->y * 32) + 0x32;
```

File: src/alph_puzzle.c [hunk 14]
```diff
@@ -0,0 +1,971 @@
...
+    data->messageFormat = MessageFormat_New_Custom(6, 16, data->heapId);
+    data->unk30 = String_New(0x80, data->heapId);
+    
+    data->unk34 = NewString_ReadMsgData(data->msgData, 0);
+    
+    for (int i = 0; i < 4; i++) {
+        data->unk38[i] = NewString_ReadMsgData(data->msgData, i + 1);
+    }
+    
+    data->unk48[0] = NewString_ReadMsgData(data->msgData, 5);
```
Comment (tgsm) [inline src/alph_puzzle.c:701]:
```suggestion
    data->unk48[0] = NewString_ReadMsgData(data->msgData, msg_0002_00005);
```

**Review by tgsm — CHANGES_REQUESTED** [review]

File: src/alph_puzzle.c [hunk 15]
```diff
@@ -0,0 +1,967 @@
...
+        thunk_OamManager_ApplyAndResetBuffers();
+    }
+
+    NNS_GfdDoVramTransfer();
+    DoScheduledBgGpuUpdates(data->bgConfig);
+
+    OS_SetIrqCheckFlag(OS_IE_VBLANK);
+}
+
+typedef struct AlphaPuzzleInitTileData {
```
Comment (tgsm) [inline src/alph_puzzle.c:547]:
```suggestion
typedef struct AlphPuzzleInitTileData {
```

**Review by tgsm — CHANGES_REQUESTED** [review]
Some comments from my last review haven't been addressed yet

File: src/alph_puzzle.c (hunk 2 above)
Comment (adrienntindall) [reply src/alph_puzzle.c:953]: Do we want to do this though? I feel like having it natively as a SysTaskFunc makes more sense

**Review by tgsm — APPROVED** [review]

**Pull request merged**